### ISSUE TITLE
Hiding Module Exports

### ISSUE THREAD

**Issue by ruby0b** [ghc-proposals/ghc-proposals#567]
This idea has been brought up in the [mailing list about 8 years ago](https://mail.haskell.org/pipermail/glasgow-haskell-users/2014-October/025373.html) and got some +1s back then but I couldn't find an actual proposal anywhere.

The idea is to allow the use the `hiding` keyword in the context of module exports.
This is useful in cases where an explicit export list would be needlessly long just to hide a single function (just like import hiding saves you from long explicit imports).

It would look like this:
```hs
module A hiding (hiddenFunc, HiddenBar(..), HiddenFoo) where

data ExposedFoo = HiddenFoo Int -- exposed type but hidden constructor
data HiddenBar = HiddenBar Int  -- hidden type and constructors
hiddenFunc = undefined          -- hidden function
```
The behavior would be identical to the currently possible workaround of having a hidden module that gets reexported:
```hs
module A (module A.Internal) where
import A.Internal hiding (hiddenFunc, HiddenBar(..), HiddenFoo) -- exactly the same definitions get exposed

module A.Internal where
data ExposedFoo = HiddenFoo Int
data HiddenBar = HiddenBar Int
hiddenFunc = undefined
```

### Limitations
Just like when using import hiding, there would be no way to hide only *either* a constructor or a type if they have the same name; both would be hidden (Haskell2010 report [section 5.3.1](https://www.haskell.org/onlinereport/haskell2010/haskellch5.html#x11-1010005.3)).

**Comment by nomeata** [comment]
I think I ran into a situation where this would have been useful recently. Not sure it means the complexity is worth it, but I'm sympathetic at least.

> Just like when using import hiding, there would be no way to hide only either a constructor or a type if they have the same name

Would ExplicitNamespaces (in their current form, or a future refined form) help here?

**Comment by ruby0b** [comment]
Yes, ExplicitNamespaces does indeed solve half of the problem. One can use `hiding (type T)` which only hides the type `T` and not the constructor with the same name `T`.
Still, the other case of wanting to hide the constructor but not the type (which seems reasonably common) would still not be possible as far as I understand.

I'll edit the original post.

**Comment by ruby0b** [comment]
Oh no, I got this all wrong. I don't think ExplicitNamespaces helps here (at least in its current form).
When you use an import with `hiding (type T)` this currently also hides the constructor named `T`.
(Is this behavior intended? It seems counter-intuitive to me.)

What does work though is using PatternSynonyms to enable `hiding (pattern T)` which hides the constructor `T` but not the type.

So the only limitation is being unable to hide a type without also hiding the constructor of the same name.

**Comment by nomeata** [comment]
That sounds like something that should be fixed. There was a recent discussion about ExplicitNamespaces elsewhere and it might benefit from an overhaul - but that needs not distract us here.
